metaflow/plugins/kubernetes/kubernetes_job.py

```diff
@@ -53,12 +56,432 @@ def wrapper(*args, **kwargs):
+                            job_spec.template.spec.containers[0]
+                            .command[idx]
+                            .replace(UBF_CONTROL, UBF_CONTROL + " " + "--split-index 0")
+                        )
+
+        job_spec.template.spec.containers[0].env = job_spec.template.spec.containers[
+            0
+        ].env + [
+            client.V1EnvVar(
+                name="MASTER_ADDR",
+                value=jobset_main_addr,
+            ),
+            client.V1EnvVar(
+                name="MASTER_PORT",
+                value=str(master_port),
+            ),
+            # In the earlier implementation, we were setting `parallelism` to created
+            # `replicatedJobs`. In this implementation, we create a different copy of the
+            # job for each replicated worker. So retrieving the rank based on the Kubernetes
+            # ENV.valueFrom (metadata.annotations['batch.kubernetes.io/job-completion-index'])
+            # will not work. since job-completion-index relies on setting the `parallelism` on the
+            # jobspec level.
+            # Instead we can statically set rank based on the index of the Job.
+            # so that workers have the right rank passed down to them.
+            client.V1EnvVar(
+                name="RANK",
+                value=str(node_index),
+            ),
+            client.V1EnvVar(
+                name="WORLD_SIZE",
+                value=str(self._kwargs["num_parallel"]),
+            ),
+            client.V1EnvVar(
+                name="PYTHONUNBUFFERED",
+                value="0",
+            ),
+        ]
+
+        return jobset.models.jobset_v1alpha2_replicated_job.JobsetV1alpha2ReplicatedJob(
+            name=job_name,
+            template=client.V1JobTemplateSpec(
+                metadata=client.V1ObjectMeta(
+                    annotations=annotations,
+                    labels=labels,
+                    namespace=self._kwargs["namespace"],
+                ),
+                spec=job_spec,
+            ),
+        )
+
+    def execute(self):
+        client = self._client.get()
+        api_instance = client.CoreV1Api()
+
+        with client.ApiClient() as api_client:
+            api_instance = client.CustomObjectsApi(api_client)
+            try:
+                jobset_obj = api_instance.create_namespaced_custom_object(
+                    self._group,
+                    self._version,
+                    self._kwargs["namespace"],
+                    "jobsets",
+                    self._jobset,
+                )
+            except Exception as e:
+                raise KubernetesJobException(
+                    "Exception when calling CustomObjectsApi->create_namespaced_custom_object: %s\n"
+                    % e
+                )
+
+        job_name = self.name + "-control-0"
+        # Retry a few times since it takes time for the job to be created
+        job_creation_timeout_sec = 60
```

Comment:
I understand this was present in even in my last implementation and we can keep this as is for now. But, we should add a comment here saying that this should be either bumped or made configurable.

In sophisticated runtime envs (e.g. with Kueue), it is possible that jobs may not be admitted for a long time. The K8s job object will only get created after admission. In those cases, failing the flow may not be ideal.

Reply:
Fixed this for good.